### PR TITLE
Miscellaneous Logger fixes

### LINKED ISSUE #1336: Reading Custom Log Quantity from GSD file
## Description

It appears that I cannot read custom log quantities from a gsd file, because they are not being written to the gsd.

## Script

To run the simulation ...
```python
import hoomd


def make_snapshot():
    snap = hoomd.Snapshot()
    snap.particles.N = 1
    snap.particles.position[:] = [[0, 0, 0]]
    snap.particles.types = ['A']
    snap.particles.typeid[:] = [0]
    return snap


class LogSingleValue(hoomd.custom.Action):

    def __init__(self):
        self._value = 5

    @hoomd.logging.log
    def value(self):
        return self._value

    def act(self):
        pass


dev = hoomd.device.CPU()
sim = hoomd.Simulation(dev)
sim.create_state_from_snapshot(make_snapshot())

value_action = LogSingleValue()
log = hoomd.logging.Logger()
log.add(value_action, quantities=['value'])

gsd_writer = hoomd.write.GSD(trigger=hoomd.trigger.Periodic(1),
                             filename='sim.gsd',
                             mode='wb',
                             log=log)
sim.operations.add(gsd_writer)

sim.run(2)
```

To try and read from the gsd ...
``` python
import gsd.hoomd

traj = gsd.hoomd.open('sim.gsd')
frame = traj[0]
value = frame.log['LogSingleValue/value']
print(value)
```

## Output

``` bash
Traceback (most recent call last):
  File "read_gsd.py", line 5, in <module>
    value = frame.log['LogSingleValue/value']
KeyError: 'LogSingleValue/value'
```

## Expected output

``` bash
5
```

## Configuration

Platform:
- Mac
- CPU

Installation method:
- Compiled from source

Versions:
- Python version: 3.7.12
- **HOOMD-blue** version: 3.2.0
- gsd version: 2.4.2

## Developer

I could have a look at this when I get the time, but @b-butler implemented the custom actions and logging subsystem, so he would probably be able to fix it faster than I could.

### PR DESCRIPTION
**Pull request by b-butler**
## Description

Removes "__main__" from custom action loggables defined in a script #1336, and fixes error in the iteration of `hoomd.util._NamespaceDict`.

## Motivation and context

Resolves #1336

## How has this been tested?

A test for `__iter__` has been added in `test_logging.py`; however, the main change is somewhat difficult to test. We test the `_NamespaceFilter` class through the logging infrastructure, but testing the removal of `__main__` is not easy to do in pytest. @tommy-waltmann could you test this locally? If desired I can add tests for the internal `_NamespaceFilter` class.

## Change log

```
Changed: Removes ``"__main__"`` from some user custom action logging namespaces.
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.md).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

**Comment by tommy-waltmann** [comment]
This does not appear to fix the issue locally. I am still seeing the `__main__` namespace before everything in `frame.log`

**Comment by tommy-waltmann** [comment]
> This does not appear to fix the issue locally. I am still seeing the `__main__` namespace before everything in `frame.log`

This is incorrect, I forgot to switch branches before trying again. After trying again I get the following error:

``` bash
Traceback (most recent call last):
  File "run_sim.py", line 13, in <module>
    class LogSingleValue(hoomd.custom.Action):
  File "/Users/tomwalt/hoomd-builds/fix-log-iter-namespace/hoomd/custom/custom_action.py", line 16, in __init__
    Loggable.__init__(cls, name, base, dct)
  File "/Users/tomwalt/hoomd-builds/fix-log-iter-namespace/hoomd/logging.py", line 314, in __init__
    log_dict.update(Loggable._get_current_cls_loggables(cls))
  File "/Users/tomwalt/hoomd-builds/fix-log-iter-namespace/hoomd/logging.py", line 353, in _get_current_cls_loggables
    entry.default)
  File "/Users/tomwalt/hoomd-builds/fix-log-iter-namespace/hoomd/logging.py", line 233, in __init__
    self.update_cls(cls)
  File "/Users/tomwalt/hoomd-builds/fix-log-iter-namespace/hoomd/logging.py", line 271, in update_cls
    self.namespace = self._generate_namespace(cls)
  File "/Users/tomwalt/hoomd-builds/fix-log-iter-namespace/hoomd/logging.py", line 280, in _generate_namespace
    return tuple(cls.namespace_filter(ns, ns[0] == "hoomd")) + (cls_name,)
  File "/Users/tomwalt/hoomd-builds/fix-log-iter-namespace/hoomd/logging.py", line 190, in __call__
    if name not in self.non_native_skip:
AttributeError: '_NamespaceFilter' object has no attribute 'non_native_skip'
```

**Comment by b-butler** [comment]
Thanks Tommy. I will continue to work on this. I will write a local test script.

**Comment by b-butler** [comment]
I just tested this, with the following script
``` python
import hoomd

class A(hoomd.custom.Action):
    @hoomd.logging.log
    def b(self):
        return 42

    def act(self, timestep):
        pass


sim = hoomd.Simulation(hoomd.device.CPU())
snapshot = hoomd.Snapshot()
snapshot.configuration.box = hoomd.Box.cube(1)
snapshot.particles.N = 0
sim.create_state_from_snapshot(snapshot)
log = hoomd.logging.Logger(["scalar"])
log += A()
print(list(log))
# [("A", "b")]
```

**Comment by b-butler** [comment]
@tommy-waltmann Yeah, I noticed the error and fixed it, so the branch should work now.

**Comment by tommy-waltmann** [comment]
Just tested locally and its working now.

File: hoomd/logging.py [hunk 1]
```diff
@@ -135,14 +135,16 @@ class _NamespaceFilter:
...
-        remove_names (set[str]): A set of names which to remove for the logging
-            namespace whenever encountered.
-        base_names (set[str]): A set of names which indicate that the next
-            encountered name in the string should be skipped. For example, if a
-            module hierarchy went like ``project.foo.bar.Bar`` and ``foo``
+        remove_names (set[str], optional): A set of names which to remove for
+            the logging namespace whenever encountered.
+        base_names (set[str], optional): A set of names which indicate that the
+            next encountered name in the string should be skipped. For example,
+            if a module hierarchy went like ``project.foo.bar.Bar`` and ``foo``
```
Comment (tommy-waltmann) [inline hoomd/logging.py]:
```suggestion
            if a module hierarchy was structured as ``project.foo.bar.Bar`` and ``foo``
```

**Review by tommy-waltmann — COMMENTED** [review]
Is it possible to add a test which does what we've been testing locally: A custom action logging to a gsd file, making sure the unwanted prefix namespace is removed?

File: hoomd/logging.py (hunk 1 above)
Comment (b-butler) [reply hoomd/logging.py]: Nope not using pytest, or a lot of custom code to test this behavior. I could test  `_NamespaceFilter` by itself.

**Comment by b-butler** [comment]
@tommy-waltmann I tested `hoomd.logging._NamespaceFilter` which is the only way to tangentially test this behavior without resorting to running scripts through subprocesses like signac-flow does.

**Review by tommy-waltmann — APPROVED** [review]
Thanks for updating this @b-butler , I think the namespaces for custom logger actions will make a lot more sense going forward.

**Comment by joaander** [comment]
@yuanzhou0827 what comments do you have on this pull request?

**Comment by joaander** [comment]
@b-butler please resolve the conflicts.

**Comment by yuanzhou0827** [comment]
I tested locally and it LGTM！

**Review by yuanzhou0827 — APPROVED** [review]

**Pull request merged**